### ISSUE TITLE
Feature Discussion: Speeding up Initial Sync when using `native_osx`

### ISSUE THREAD

**Issue by rwilliams** [EugenMayer/docker-sync#383]
By combing the `unison` and `native_osx` sync strategies there might be a way to have very fast initial sync and continue to enjoy the reduced overhead of the current `native_osx` strategy.

#### 2. Use Unison for initial copy from local filesystem to /host_sync. Here /host_sync is a real dir on the sync-container
This should be pretty fast and allow for excludes, etc.

#### 3. Do initial sync of /host_sync and /app_sync with unison. 
This should be very fast as there isn't a volume share and should build the indexes for unison file watcher to start quickly.  

#### 6. start unison watcher for /host_sync <-> /app_sync on sync-container
This is where things get interesting. Will unison freak out when this happens or will it pretend nothing happened and use the indexes already created? I'm trying to put together a proof on concept on this one.

Thoughts?

**Comment by EugenMayer** [comment]
Pros
 1) at least we will not need unox
 2) probably speed ( i kinda doubt that, did you benchmark it?)

Cons:
 a) This would mean we have a dependency on unison on OSX again
 b) brew dependency again
 c) since the sync src folder is on the host, we do no longer speed up the second start of unison, since the index is on the osx host
 d) complexity - we basically do unison-> unison for intial and osxfs->unison during runtime, code complexity is def. higher
 e) when we start the second time, unison has to compare host_sync and app_sync using a deep algo with fingerprints, since it does absolutely not know that we already synced before and it does not have the sync protocol/index, since it stays on the host sync

Answers:
Unison will not freak out, since we would not sync on /host_sync but rather to /app_sync - syncing to host_sync will be useless since if we runtime start and host mount that folder would be overlayed

---

Bottom line would be, i would not like to discourage you, but my gut feeling tells me, that it is not worth it.

i think the speedup will not be worth all the cons, especially if we consider 2 things, not only the speed for the initial sync but rather the time it takes to start into runtime and the CPU-usage drops . (unison started and indexed). I think, calling that stage 2, it will probably even end up being worse in overall time.

My suggestion would be, not speed up the initial sync, but rather make the initial sync reusable a lot more often. Currently, i rather never clean the sync, i stop it and only clean the app-stack, but then just start the sync - which does not force me into a re sync.

so instead of probably speed up the "1-time task" which really does happen to you probably one time in the morning and gives you a 60 seconds speedup, rather ensure that we can reuse that sync the whole day, even if we need to flush the app-stack.

I really appreciate all your thoughts and cross-thinking!

**Comment by rwilliams** [comment]
Yeah it's probably way overkill. I think when you first brought up using Unison for the initial precopy for `native_osx` I was expecting the `native_osx` strategy to start as fast as `unison`. 

#### Pros:
2. I think there would be a significant speed gain on the initial sync. It should run almost as fast as the pure `unison` strategy. The only difference between this and pure `unison` on the initial sync would be an extra step of syncing from /host_sync to /app_sync and when this is done without a volume share it runs blazing fast.

#### Cons:
a,b: We could try running unison from a docker container instead of being dependent on homebrew.

c,e: Step 3 should build an index on the app-sync container. The key is will unison re-use this index when we switch out the real dir /host-sync with the volume share /host-sync (steps 4 and 5). They should be essentially identical as we basically just made two clones of our project source.

#### TLDR
At the end of the day it was a pretty fun mental exercise but probably not worth it. Even if my best cause assumptions above work it's probably not worth the added complexity to save a bit of time on a task that you won't run very often. The benefits of `docker-sync` are still going to be awesome.

**Comment by EugenMayer** [comment]
@rwilliams 
a,b) .. wont work..if we do so, you will need to mount the host src into that container..and guess how fast things will run then? :)

c/e) not possible, since there is anything inside, for atime to absolute pathes i guess and its stored somewhere in ~/.unison of the users. Reusing that index is probably completely impossible

---

in the end, we can do something totally different in the very new future, but you did not ask me to share it :)

Think about using :delegated when we mount for the initial sync. Guess what happens? :)
There is also something else. We could use `docker cp` to /host_sync and then unison to /app_sync - but i did not measure the speedup. I guess waiting for delegated makes this more elegant to implement

**Comment by rwilliams** [comment]
@EugenMayer I guess I can continue suffering through a whole 9 seconds of precopy (with cached enabled) while you cook up that magic :)

**Comment by EugenMayer** [comment]
@rwilliams may i then close this ticket for now?

**Comment by rwilliams** [comment]
Go for it
